src/typing/typingReducer.js

```diff
@@ -14,6 +14,47 @@ import { NULL_OBJECT } from '../nullObjects';
 
 const initialState: TypingState = NULL_OBJECT;
 
+/**
+ * Typing state sans `senderId` in state[normalizedRecipients].userIds
+ *
+ * With `time` in state[normalizedRecipients] set to the provided
+ * value.
+ *
+ * PRIVATE, exported only for tests.
+ */
+export const stateWithoutSenderId = (
+  state: TypingState,
+  args: {|
+    normalizedRecipients: string,
+    senderId: number,
+    time: number,
+  |},
+) => {
+  const { normalizedRecipients, senderId, time } = args;
+
+  if (state[normalizedRecipients] === undefined) {
+    return state;
+  }
+
+  const newTypingUsers = state[normalizedRecipients].userIds.filter(userId => userId !== senderId);
+
+  if (newTypingUsers.length > 0) {
+    return {
+      ...state,
+      [normalizedRecipients]: {
+        ...state[normalizedRecipients],
+        time,
```

Comment:
Another thing that occurs to me on looking at this code, and in light of your later commits about TYPING_CLIENT_TIMER_EXPIRED:

It doesn't seem to make a lot of sense to bump `time` here. (Which we do both before and after this commit.) Effectively we're saying that if people A and B were typing, and then B stops typing... that will push back the time at which we should decide A isn't typing any more if we haven't heard another start from them.

And in fact because we only have the one "expired" action dispatched per start, we'll ignore that action and just keep thinking they're still typing forever.

In order to accurately track all this, I think what we really need is a separate time per sender -- namely the time we last heard of a typing start from them.